### PR TITLE
examples : add debug utility/example

### PR DESCRIPTION
**Pull request by danbev**
This commit introduces a new example named llama-debug which is a utility that is intended to be used to assist with developing/debugging a converted model.

The motivation for this utilitiy is to assist in model conversion work to verify that the model produces the expected outputs. It is intended to replace logits.cpp in examples/model-conversion.

Example usage:
```console
./build/bin/llama-debug \
    -m models/Qwen2.5-0.5B-Instruct.gguf \
    --prompt "Hello, my name is" \
    --save-logits
...
Model add_bos: false
Input prompt: "Hello, my name is"
Token ids (5):
Hello(9707) ,(11)  my(847)  name(829)  is(374)
Data saved to data/llamacpp-Qwen2.5-0.5B-Instruct.bin
Data saved to data/llamacpp-Qwen2.5-0.5B-Instruct.txt
Prompt saved to data/llamacpp-Qwen2.5-0.5B-Instruct-prompt.txt
Tokens saved to data/llamacpp-Qwen2.5-0.5B-Instruct-tokens.bin
```

For more details about the options available for this example, please refer to examples/debug/README.md.

----
This was suggested/discussed in the following pr: https://github.com/ggml-org/llama.cpp/pull/18281#issuecomment-3681892592

### CODE REVIEW

**Comment by pwilkin** [comment]
Does this integrate the `llama-eval-callback` functionality? Seems like it might be a good idea to phase that one out.

**Comment by danbev** [comment]
> Does this integrate the `llama-eval-callback` functionality?

Yes, this uses the eval-callback feature similar to `llama-eval-callback`, and we could indeed phase out llama-eval-callback  and use this as the reference for that functionality instead.

File: examples/debug/debug.cpp [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+
+    output_data(llama_context * ctx, const llama_model * model, const common_params & params) {
+        const llama_vocab * vocab = llama_model_get_vocab(model);
+        const bool add_bos = llama_vocab_get_add_bos(vocab);
+
+        tokens = common_tokenize(ctx, params.prompt, add_bos);
+        prompt = params.prompt;
+
+        if (params.embedding) {
+            const int  n_embd          = llama_model_n_embd(model);
```
Comment (ggerganov) [inline examples/debug/debug.cpp]: This probably should become `llama_model_n_embd_out()`

**Review by ggerganov — APPROVED** [review]

File: examples/debug/debug.cpp (hunk 1 above)
Comment (danbev) [reply examples/debug/debug.cpp]: Yep, I'll update this with the changes from https://github.com/ggml-org/llama.cpp/pull/18607 tomorrow :+1:

File: examples/debug/debug.cpp [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+        save_output_data(output, model_name, params.logits_output_dir);
+    }
+
+    return true;
+}
+
+int main(int argc, char ** argv) {
+    common_params params;
+
+    if (!common_params_parse(argc, argv, params, LLAMA_EXAMPLE_DEBUG)) {
```
Comment (ngxson) [inline examples/debug/debug.cpp]: It's recommended to add an example usage via `print_usage`, an example can be found in `simple.cpp`

File: examples/debug/debug.cpp [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+    LOG("\n");
+}
+
+static bool run(llama_context * ctx, const common_params & params) {
+    const llama_model * model = llama_get_model(ctx);
+    const llama_vocab * vocab = llama_model_get_vocab(model);
+
+    const bool add_bos = llama_vocab_get_add_bos(vocab);
+
+    std::vector<llama_token> tokens = common_tokenize(ctx, params.prompt, add_bos);
```
Comment (ngxson) [inline examples/debug/debug.cpp:357]: can we allow entering raw tokens? maybe reuse the `-p` option:

```
-p "hello" --> treat as text input
-p "tokens:12,34" --> treat as 2 input tokens
```

we can skip adding bos/eos in such case. this use case is mostly to isolate testing inference and tokenizer

**Review by ngxson — APPROVED** [review]

File: examples/debug/debug.cpp (hunk 2 above)
Comment (danbev) [reply examples/debug/debug.cpp]: Thanks, I'll take a look and add this.

File: examples/debug/debug.cpp (hunk 3 above)
Comment (danbev) [reply examples/debug/debug.cpp:357]: I'll take a look and see what would be involved.

Comment (danbev) [reply examples/debug/debug.cpp:357]: @ngxson I just wanted to ask if you saw that we recently (in this PR) added support for also saving the prompt and token ids when using the `--save-logits` flag. 
For example, running the following command:
```console
$ ./build/bin/llama-debug -m ../llama.cpp/models/Qwen2.5-0.5B-Instruct.gguf -p "Hello world today" --save-logits
...
Model add_bos: false
Input prompt: "Hello world today"
Token ids (3):
Hello(9707)  world(1879)  today(3351) 
Data saved to data/llamacpp-Qwen2.5-0.5B-Instruct.bin
Data saved to data/llamacpp-Qwen2.5-0.5B-Instruct.txt
Prompt saved to data/llamacpp-Qwen2.5-0.5B-Instruct-prompt.txt
Tokens saved to data/llamacpp-Qwen2.5-0.5B-Instruct-tokens.bin
```
The `-prompt.txt` file would then contain the following information:
```console
(venv) $ cat data/llamacpp-Qwen2.5-0.5B-Instruct-prompt.txt 
prompt: Hello world today
n_tokens: 3
token ids: 9707, 1879, 3351
```
And the `-tokens.bin` file contains just the token ids which we then use in the first step of the logits/embeddings verification to make sure that we are using the same inputs.

Would this be enough or do you think that being able to specify the actual input token ids would useful to have?

Comment (danbev) [reply examples/debug/debug.cpp:357]: I'm going to merge this to continue on #18658, but I'm happy to add this in new PR if needed.

File: examples/debug/debug.cpp [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+
+            float * embeddings;
+            if (pooling_enabled) {
+                embeddings = llama_get_embeddings_seq(ctx, 0);
+                storage.resize(n_embeddings);
+                common_embd_normalize(embeddings, storage.data(), n_embeddings, params.embd_normalize);
+                embeddings = storage.data();
+            } else {
+                embeddings = llama_get_embeddings(ctx);
+            }
```
Comment (ggerganov) [inline examples/debug/debug.cpp:89]: @danbev I'm not very familiar with the embedding use cases, but my guess would be that we also want to be able to normalize the embeddings even if there is no pooling. Any reason to disable this option here?

File: examples/debug/debug.cpp [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+            const int  n_embd          = llama_model_n_embd_out(model);
+            const bool pooling_enabled = llama_pooling_type(ctx) != LLAMA_POOLING_TYPE_NONE;
+            const int  n_embd_count    = pooling_enabled ? 1 : tokens.size();
+            const int  n_embeddings    = n_embd * n_embd_count;
+
+            float * embeddings;
+            if (pooling_enabled) {
+                embeddings = llama_get_embeddings_seq(ctx, 0);
+                storage.resize(n_embeddings);
+                common_embd_normalize(embeddings, storage.data(), n_embeddings, params.embd_normalize);
```
Comment (ggerganov) [inline examples/debug/debug.cpp:85]: Instead of normalizing the concated embeddings, we should normalize token-by-token.

File: examples/debug/debug.cpp (hunk 4 above)
Comment (danbev) [reply examples/debug/debug.cpp:89]: > Any reason to disable this option here?

No, this was a mistake on my part carried over from logits.cpp. It should be possible to normalize the embeddings regardless if pooling is enabled or not. I'll take a look at fixing this now. Thanks

Comment (danbev) [reply examples/debug/debug.cpp:89]: I've address this in https://github.com/ggml-org/llama.cpp/pull/18692/commits/f31c38dd19d012750146945ed930637d6a1050f2.

File: examples/debug/debug.cpp (hunk 5 above)
Comment (danbev) [reply examples/debug/debug.cpp:85]: I've addressed this in https://github.com/ggml-org/llama.cpp/pull/18692/commits/f31c38dd19d012750146945ed930637d6a1050f2.
